### ISSUE TITLE
foreign exception error

### ISSUE THREAD

**Issue by ralyodio** [Atome-FE/llama-node#76]
llama.cpp: loading model from /home/ubuntu/models/ggml-gpt4all-j-v1.3-groovy.bin
fatal runtime error: Rust cannot catch foreign exceptions
Aborted2023-05-18T11:49:39.270396Z  INFO surrealdb::net: SIGTERM received. Start graceful shutdown...    
2023-05-18T11:49:39.270450Z  INFO surrealdb::net: Shutdown complete. Bye!    

==> /var/log/descriptive-web.log <==
Now using nod

**Comment by ralyodio** [comment]
again catch and throw but we shouldn't be crashing or exiting when this happens.

**Comment by hlhr202** [comment]
This is not a llama based model, so it cannot be used on llama.cpp backend.
According to its [hugging face page](https://huggingface.co/nomic-ai/gpt4all-j), the model is based on [GPT-J](https://huggingface.co/EleutherAI/gpt-j-6B)
I m not sure if llm-rs backend supports this or not, but llama.cpp definitely not.

**Comment by ralyodio** [comment]
yeah should it crash though?? log an error I understand. but crash??

**Comment by hlhr202** [comment]
> yeah should it crash though?? log an error I understand. but crash??
> 
> i have this code wrapped in a try/catch from my side and it isn't catching.

this error is not catchable at current stage. or you can open issue on llama.cpp github space. loading model file is unsafe for rust side and i can only get a nullprt when loaded failed. but in this case, i cant even get any response from the low level side.

generally, you shouldnt use this model with llama.cpp at all. the termination and error is not thrown to upper level. its more like playing fire.

**Comment by ralyodio** [comment]
in my use case i have an api that uses this library, it allows people to pick from a list of models. Its possible I have models that aren't supported, so ideally it wouldn't take down my entire backend if someone picked an incompatible file.

you can't do try/catch around file loading?

**Comment by hlhr202** [comment]
> 

yea. basically this is very early stage library and not production ready. every release i v tagged with pre release. i admit this is essentially a toy app. and my upper stream llama.cpp/llm-rs even updating every few hours. how can i promise that everything just work in purpose?

**Comment by hlhr202** [comment]
or maybe you can do something with it maybe find a way to fix. i wont suggest you use it in any production scenario or you have to keep the dangerous away from user input.

**Comment by hlhr202** [comment]
i will leave this temporily as bug till i find a way to fix

**Comment by hlhr202** [comment]
double confirmed this error should be handled in llama.cpp

**Comment by linonetwo** [comment]
This fatal error will crash the whole app even try-catch + worker_thread is used. Better if you can just throw a JS error (sub-class so we can switch-case the error) instead!

**Comment by HolmesDomain** [comment]
I have a similar issue as well. Hope it gets fixed in llama.cpp or wherever.
